aten/src/ATen/native/TensorDimApply.h

```diff
@@ -0,0 +1,55 @@
+#include <ATen/ATen.h>
+
+namespace at {
+  namespace native {
+    //input tensors are non-zero dim and non-empty
+    template<typename T1, typename T2, typename Operation>
+    void tensor_dim_apply3(const Tensor& self, Tensor& values, Tensor& indices, int64_t dim, Operation op) {
+      int ndims = self.dim();
+      int tensor_dim_apply_has_finished = 0;
+      int tensor_dim_apply_i;
+      int64_t tensor_dim_apply_counter[ndims];
+      for(tensor_dim_apply_i = 0; tensor_dim_apply_i < ndims; tensor_dim_apply_i++)
+        tensor_dim_apply_counter[tensor_dim_apply_i] = 0;
+      T1* self_data = self.data_ptr<T1>();
+      T2* values_data = values.data_ptr<T2>();
+      int64_t* indices_data = indices.data_ptr<int64_t>();
```

Comment:
It's better not to hardcode indices type - what if we decide to support int32_t or byte indices for small arrays?

Reply:
good point! will add a template type